File: day13/typescript/day13.ts
import { readFile } from 'fs';
import { promisify } from 'util';

import * as R from 'ramda';

interface Layer {
    depth: number;
    range: number;
    scanner: number;
}

type Firewall = { [depth: number]: Layer };

const parse = R.compose<string[], Firewall[], Firewall>(
    R.mergeAll,
    R.map(
        R.compose<string, string[], number[], Firewall>(
            ([d, r]) => ({ [d]: { depth: d, range: r, scanner: 0 } }),
            R.map(parseInt),
            R.split(': '),
        ),
    ),
);

const stepper = (fn: (s: number, k: number) => number) => R.mapObjIndexed(
    ({ scanner, range, depth }, key, obj) => ({
        depth,
        range,
        scanner: fn(scanner, parseInt(key, 10)) % (range * 2 - 2),
    }),
);

const step = stepper((s, k) => s + 1);
const sim = stepper(R.add);

const part1 = (firewall: Firewall) => R.reduce<Layer, number>(
    (acc, { depth, range }) => acc + depth * range,
    0,
    R.filter(l => l.scanner === 0, R.values(sim(firewall))),
);

const part2 = (firewall: Firewall) => {
    for (let i = 0;; ++i) {
        const sim2 = stepper((s, k) => s + k + i);
        if (R.all(({ scanner }) => scanner !== 0, R.values(sim2(firewall)))) {
            return i;
        }
    }
};

(async () => {
    const input = await promisify(readFile)('day13/input.txt', 'utf8');
    const firewall = parse(R.split('\r\n', input));

    console.log(part1(firewall));
    console.log(part2(firewall));
})();

File: day13/haskell/day13.hs
import Data.Char (isDigit)
import Data.List (findIndex)

type Layer = (Int, Int, Int)
type Firewall = [Layer]

parse :: String -> Firewall
parse = map (f . map (read . filter isDigit) . words) . lines
    where f [d, r] = (d, r, 0)

stepper :: Int -> Firewall -> Firewall
stepper i = map step
    where step (d, r, s) = (d, r, (s + d + i) `mod` (r * 2 - 2))

hit :: Layer -> Bool
hit (_, _, s) = s == 0

part1 :: Firewall -> Int
part1 = sum . map sev . filter hit . stepper 0
    where sev (d, r, _) = d * r

part2 :: Firewall -> Maybe Int
part2 f = fmap (+ 1) . findIndex (not . any hit . ($ f)) . map stepper $ [1..]

main :: IO ()
main = do
    firewall <- parse <$> readFile "day13/input.txt"
    print . part1 $ firewall
    print . part2 $ firewall